package/kedro_viz/api/rest/responses/run_events.py

```diff
@@ -0,0 +1,447 @@
+    error: Optional[Union[str, DatasetErrorInfo]] = None
+
+
+class PipelineInfo(BaseModel):
+    """Information about the pipeline run."""
+
+    run_id: str = "default-run-id"
+    start_time: Optional[str] = None
+    end_time: Optional[str] = None
+    total_duration_sec: float = 0.0
+    status: str = PipelineStatus.COMPLETED
+    error: Optional[str] = None
+
+
+class StructuredRunStatusAPIResponse(BaseModel):
+    """Format for structured run status endpoint response."""
+
+    nodes: Dict[str, NodeInfo] = Field(default_factory=dict)
+    datasets: Dict[str, DatasetInfo] = Field(default_factory=dict)
+    pipeline: PipelineInfo = Field(default_factory=PipelineInfo)
+
+
+def _create_or_update_dataset_info(
+    datasets: Dict[str, DatasetInfo],
+    dataset_id: str,
+    dataset_name: str,
+    size_bytes: Optional[int],
+    status: str,
+    overwrite_size: bool = False,
+) -> None:
+    """Create or update DatasetInfo, with optional size overwrite control.
+
+    Args:
+        datasets: Dictionary of existing dataset info objects
+        dataset_id: Unique identifier for the dataset
+        dataset_name: Name of the dataset
+        size_bytes: Size of the dataset in bytes
+        status: Status of the dataset
+        overwrite_size: Whether to overwrite existing size information
+    """
+    status_enum = convert_status_to_enum(status, DatasetStatus.AVAILABLE)
+
+    if dataset_id not in datasets:
+        datasets[dataset_id] = DatasetInfo(
+            name=dataset_name,
+            size_bytes=size_bytes or 0,
+            status=status_enum,
+        )
+        return
+
+    # Update existing dataset info
+    dataset_info = datasets[dataset_id]
+    if not dataset_info.name and dataset_name:
+        dataset_info.name = dataset_name
+
+    if size_bytes is not None and (overwrite_size or dataset_info.size_bytes == 0):
+        dataset_info.size_bytes = size_bytes
+
+    dataset_info.status = status_enum
+
+
+def _extract_pipeline_timing_and_status(
+    events: List[Dict[str, Any]], pipeline_info: PipelineInfo
+) -> None:
+    """Extract pipeline start time, end time, and status from events.
+
+    Args:
+        events: List of event dictionaries
+        pipeline_info: Pipeline info object to update
+    """
+    # Find pipeline start event
+    start_event = next(
+        (
+            event
+            for event in events
+            if (
+                event.get("event") == EventType.BEFORE_PIPELINE_RUN
+                and event.get("timestamp")
+            )
+        ),
+        None,
+    )
+    if start_event:
+        pipeline_info.start_time = start_event["timestamp"]
+
+    # Find pipeline end event (last pipeline-related event)
+    end_events = [
+        event
+        for event in reversed(events)
+        if event.get("event")
+        in {EventType.AFTER_PIPELINE_RUN, EventType.ON_PIPELINE_ERROR}
+    ]
+
+    if end_events:
+        end_event = end_events[0]
```

Comment:
Can there ever be more than 1 event? And why store it in a list if that's not possible?